File: Lab 1/q1.cpp
#include <iostream>
#include <fstream>
#include <bitset>

std::string decToBin(int n) {
    if (n == 0)
        return "0000000000000000";
    else
        return decToBin(n / 2) + std::to_string(n % 2);
}

int main(int argc, char* argv[]) {
    if (argc != 4) {
        std::cout << "Usage: " << argv[0] << " <n> <input_file_name> <output_file_name>\n";
        return 1;
    }

    int n = std::atoi(argv[1]);
    if (n <= 0) {
        std::cout << "Invalid value of 'n'. Please provide a positive integer.\n";
        return 1;
    }

    std::ifstream inputFile(argv[2]);
    std::ofstream outputFile(argv[3]);

    if (!inputFile || !outputFile) {
        std::cout << "Error opening files.\n";
        return 1;
    }

    int num;
    for (int i = 0; i < n; ++i) {
        if (inputFile >> num) {
            std::string binary = decToBin(num);
            outputFile << "The binary equivalent of " << num << " is " << binary << std::endl;
            std::cout << "The binary equivalent of " << num << " is " << binary << std::endl;
        } else {
            std::cout << "Reached end of input file.\n";
            break;
        }
    }

    inputFile.close();
    outputFile.close();

    return 0;
}

File: Lab 1/q1.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;

public class q1 {

    public static void main(String[] args) {
        if (args.length != 3) {
            System.out.println("Usage: java q1 <n> <input_file> <output_file>");
            return;
        }

        int n = Integer.parseInt(args[0]);
        String inputFileName = args[1];
        String outputFileName = args[2];

        try {
            BufferedReader reader = new BufferedReader(new FileReader(inputFileName));
            BufferedWriter writer = new BufferedWriter(new FileWriter(outputFileName));
            String line;
            while ((line = reader.readLine()) != null && n > 0) {
                int decimal = Integer.parseInt(line);
                String binary = convertToBinary(decimal, 16); // You can change the second argument to represent the number of bits for the binary representation
                writer.write("The binary equivalent of " + decimal + " is " + binary + "\n");
                n--;
            }
            reader.close();
            writer.close();
        } catch (IOException e) {
            System.out.println("Error reading/writing files: " + e.getMessage());
        }
    }

    private static String convertToBinary(int decimal, int numBits) {
        if (numBits == 0) {
            return "";
        } else {
            int bit = decimal % 2;
            decimal /= 2;
            String partialBinary = convertToBinary(decimal, numBits - 1);
            return partialBinary + bit;
        }
    }
}
